### ISSUE TITLE
compile time hardening flags, canary, PIE

### ISSUE THREAD

**Issue by adrelanos** [GrapheneOS/hardened_malloc#84]
This project is over my head. During considerations of packaging it, I randomly found something.

```
checksec --file libhardened_malloc.so 
RELRO           STACK CANARY      NX            PIE             RPATH      RUNPATH      FILE
Full RELRO      No canary found   NX enabled    DSO             No RPATH   No RUNPATH   libhardened_malloc.so
```

```
hardening-check libhardened_malloc.so 
libhardened_malloc.so:
 Position Independent Executable: no, regular shared library (ignored)
 Stack protected: no, not found!
 Fortify Source functions: no, only unprotected functions found!
 Read-only relocations: yes
 Immediate binding: yes
```

You tell me if worth fixing or pointless for this package. In case it's pointless, just close this ticket.

**Comment by thestinger** [comment]
> This project is over my head. During considerations of packaging it, I randomly found something.

You should be aware that the scripts you're using to check for hardening can be misleading due to false positives and lack support for checking for modern mitigations like type-based CFI and SafeStack. The check for SSP (i.e. weak stack canaries for detecting some linear overflows) assumes that there was a function triggered by the heuristic, and is ignorant of the fact that there are better approaches than SSP. The FORTIFY_SOURCE check similarly misses that the feature is designed to avoid adding a check in cases that are statically verified as correct. It's also designed around the low quality glibc implementation of FORTIFY_SOURCE which is unnecessarily incompatible with Clang and doesn't check for buffer read overflows, which also means fewer functions have fortified versions.

The traditional mitigations are enabled by default in any modern toolchain / OS taking basic security measures. Here are the results on a non-security-focused Linux distribution (Arch Linux):

```
% checksec --file libhardened_malloc.so
RELRO           STACK CANARY      NX            PIE             RPATH      RUNPATH	Symbols		FORTIFY	Fortified	Fortifiable  FILE
Full RELRO      Canary found      NX enabled    DSO             No RPATH   No RUNPATH   178 Symbols     Yes	0		6	libhardened_malloc.so
```

You'll definitely want to use hardened_malloc on an OS where these things are all enabled by default and less ancient mitigations like SafeStack and type-based CFI are at least in the process of being integrated.

As it mentions in the README, there's libc hardening meant to go along with it, but there's a lot more to it than that. The README is hinting at the fact that this shouldn't be treated as a magic bullet but rather one small part of a systemic approach to security. If your toolchain / OS isn't enabling basic early 2000s mitigations by default, you should probably address that first. There are many other things to work on alongside this too and possibly before it. For example, in a traditional desktop operating system with a non-existent security model for applications, this is definitely not what I would prioritize.

>  Position Independent Executable: no, regular shared library (ignored)

A dynamic library cannot be a Position Independent Executable (PIE). There is not executable produced by the build process. The code is fully position independent (PIC) and the shared object (DSO) is inherently position independent.

It's also really the responsibility of your toolchain and OS to use PIE by default. If it doesn't produce position independent executables by default in 2019, it's not a sane toolchain / OS.

>  Stack protected: no, not found!

Similarly, this is really the responsibility for your toolchain and OS. SSP doesn't really offer anything of value here because what's needed is protection against *external* code attacking the data on the stack and SSP doesn't provide it.

> You tell me if worth fixing or pointless for this package.

I don't really see anything to fix in this project but rather flaws with the toolchain that you're using to build it.

**Comment by thestinger** [comment]
> This project is over my head. During considerations of packaging it, I randomly found something.

You should be aware that the scripts you're using to check for hardening can be misleading due to false positives and lack support for checking for modern mitigations like type-based CFI and SafeStack. The check for SSP (i.e. weak stack canaries for detecting some linear overflows) assumes that there was a function triggered by the heuristic, and is ignorant of the fact that there are better approaches than SSP. The FORTIFY_SOURCE check similarly misses that the feature is designed to avoid adding a check in cases that are statically verified as correct. It's also designed around the low quality glibc implementation of FORTIFY_SOURCE which is unnecessarily incompatible with Clang and doesn't check for buffer read overflows, which also means fewer functions have fortified versions.

The traditional mitigations (SSP with -fstack-protector-strong, full RELRO, PIE) are enabled by default in any modern toolchain / OS taking basic security measures. Here are the results on a non-security-focused Linux distribution (Arch Linux):

```
% checksec --file libhardened_malloc.so
RELRO           STACK CANARY      NX            PIE             RPATH      RUNPATH	Symbols		FORTIFY	Fortified	Fortifiable  FILE
Full RELRO      Canary found      NX enabled    DSO             No RPATH   No RUNPATH   178 Symbols     Yes	0		6	libhardened_malloc.so
```

You'll definitely want to use hardened_malloc on an OS where these things are all enabled by default and less ancient mitigations like SafeStack and type-based CFI are at least in the process of being integrated.

As it mentions in the README, there's libc hardening meant to go along with it, but there's a lot more to it than that. The README is hinting at the fact that this shouldn't be treated as a magic bullet but rather one small part of a systemic approach to security. If your toolchain / OS isn't enabling basic early 2000s mitigations by default, you should probably address that first. There are many other things to work on alongside this too and possibly before it. For example, in a traditional desktop operating system with a non-existent security model for applications, this is definitely not what I would prioritize.

>  Position Independent Executable: no, regular shared library (ignored)

A dynamic library cannot be a Position Independent Executable (PIE). There is not executable produced by the build process. The code is fully position independent (PIC) and the shared object (DSO) is inherently position independent.

It's also really the responsibility of your toolchain and OS to use PIE by default. If it doesn't produce position independent executables by default in 2019, it's not a sane toolchain / OS.

>  Stack protected: no, not found!

Similarly, this is really the responsibility for your toolchain and OS. SSP doesn't really offer anything of value here because what's needed is protection against *external* code attacking the data on the stack and SSP doesn't provide it. Similarly, there is nothing to actually protect with `_FORTIFY_SOURCE`. The successor to the old `checksec` that you're using identifiers 6 functions that could be potentially protected and enabling it results in 2/6 calls being to checked variants. However, both of those calls are dead code error output tied to reporting division by zero errors in libdivide. So, I could enable it, but it wouldn't accomplish anything, similar to SSP. Perhaps it's worth enabling due to perception but those 2 calls it can 'protect' are likely going to be wiped out by the compiler getting less terrible at optimization because it's very obvious that it never has an iteration of the initialization loop where the passed divisor could be zero but the compiler apparently can't reason about values read from a constant array.

> You tell me if worth fixing or pointless for this package.

It's worth fixing your toolchain to use SSP by default, and it's probably also missing using PIE and full RELRO by default too. Using full RELRO is even a runtime performance increase if you couple it with using `-fno-plt` as an optimization.

I don't see anything to address for hardened_malloc beyond considering overriding the default SSP configuration with `-fstack-protector-strong` but it would be about sacrificing a minuscule amount of performance to improve perception rather than it having actual security value for this case.

**Comment by thestinger** [comment]
I enabled -fstack-clash-protection in this commit, although it's a no-op for the code as it currently exists and should probably always remain one:

https://github.com/GrapheneOS/hardened_malloc/commit/77743e5a36cdda57bebd163e4b7fec72a22eb865

I added -fstack-protector-strong as an override of the default configuration in the toolchain in this commit:

https://github.com/GrapheneOS/hardened_malloc/commit/de8cf0938a2d9fe36f36da4e0af66d2592c84ed3

The commit message explains that this is expected to generally be a no-op where this project is deployed, and recommends using ShadowCallStack on arm64 instead of SSP. I can't just enable ShadowCallStack on arm64, because I can't rely on there being runtime support. The same goes for the SafeStack approach, which has no production implementation or cross-DSO support available for traditional Linux distributions, only the compiler-rt implementation that is not a very serious take on it. That really requires libc integration. ShadowCallStack is much simpler and easier to integrate, but they decided not to go ahead with supporting it for x86_64 at this time due to the inherent race involved in return checks on x86_64 since the return address on the stack is used when returning vs. having a link register on arm, etc.
